### ISSUE TITLE
missing rule: disallow implicit type conversion

### ISSUE THREAD

**Issue by remitbri** [eslint/eslint#1621]
Hi. I've tried to review each and every rule of JSHint, JSCS and ESLint to see if I could entirely "get rid of" the first two in favour of the latter one. As suggested by the owner of the @geteslint twitter account, I'm reporting the JSCS rules I've found actually not covered by ESLint. My apologies if it's a duplicate.

- "disallowImplicitTypeConversion"

The rule is set with an array of quoted types: "numeric", "boolean", "binary", "string"

E.g this is invalid

```js
x = !!y
x = +y
x = "" + y
x = ~s.indexOf(".")
```

**Comment by michaelficarra** [comment]
What is implicit about any of those? That is as explicit as it gets to me. Implicit would be

```js
if(x.length) ...
```

instead of

```js
if(x.length > 0) ...
```

**Comment by nzakas** [comment]
Let's not get hung up on naming. That's what JSCS called this setting. We should have an equivalent.

**Comment by michaelficarra** [comment]
But I don't understand its purpose. Do we just want to blindly copy all JSCS rules?

**Comment by nzakas** [comment]
If our goal is to allow people to replace JSCS with ESLint, then we need to support this. That's not blind copying, it's a feature people want.

**Comment by lo1tuma** [comment]
I agree with @michaelficarra, it is blind copying if we don’t have a precise description what this rule/feature should do.

**Comment by nzakas** [comment]
It's pretty clear to me what the rule does: disallows coercion. It flags operations that result in coercion and forces you to do things like `String(foo)` instead of `foo + "" `.

**Comment by xjamundx** [comment]
Feels like cheating but here's a starting place :)
https://github.com/jscs-dev/node-jscs/blob/master/lib/rules/disallow-implicit-type-conversion.js

**Comment by doberkofler** [comment]
+1

**Comment by mysticatea** [comment]
I will try to make this.

Would be the name `no-implicit-type-conversion`? Which or `no-implicit-cast`?

**Comment by mysticatea** [comment]
I think `no-tricky-cast` might be good.

And I thought I want to change options from the proposal: https://gist.github.com/mysticatea/120ab01b8cc11e5e36da
Thought?

**Comment by ilyavolodin** [comment]
Options look fine to me. I'm not sold on the name though... Tricky is too broad of a word. I don't really think there's anything tricky about `!!` conversion.

**Comment by michaelficarra** [comment]
@mysticatea Please not "tricky". I think `no-implicit-cast` is fine.

@ilyavolodin The rule would warn the opposite way around: when those explicit casts are not there.

**Comment by ilyavolodin** [comment]
@michaelficarra I thought original request was for flagging expression like ```var x = !!y```

**Comment by michaelficarra** [comment]
A rule like that wouldn't make much sense. How would you be expected to cast to bool? `a ? true : false`? Use `parseFloat` to cast to number? Those are just as explicit, but not as terse. I couldn't see anybody using that rule, while I'm sure many people would want the opposite: no implicit casts.

**Comment by ilyavolodin** [comment]
@michaelficarra ```Boolean(a)```? I'm not saying I would use a rule like that, but there are a lot of different coding styles out there, I'm sure somebody would want it (since that rule existing in JSCS).

**Comment by michaelficarra** [comment]
So

```js
"a" + String(b) + "c" + String(d) + "e" + String(f)
```

instead of 

```js
"a" + b + "c" + d + "e" + f
```

and

```js
!Boolean(a)
```

instead of

```js
!a
```

and

```js
Number(a) + 1
```

instead of

```js
a + 1
````

Really, I can't imagine ever seeing code like that. Anyway, my vote for name is now `type-annotated-operands` since all operands must always explicitly write down what conversion the operator will use if they are inappropriately typed. Should I open a separate issue for a rule that warns against implicit coercions?

**Comment by ilyavolodin** [comment]
@michaelficarra I agree, for me personally, this style would not be very useful, but there are a lot of users of ESLint out there. `type-annotated-operands` sounds like a good name.
Yes, please open a new issue for implicit coercion. I, however, think it will be extremely hard to create a comprehensive rule for that without duplicating most of the Flow.

**Comment by michaelficarra** [comment]
You're right, anything more complex than ensuring the `test` of an iteration statement is a comparison or a logical negation -- basically anything non-local -- would require an inference system. Oh well...

**Comment by mysticatea** [comment]
ESLint is a tool to help developers. "Implicit type conversion" also is a feature to help developers, most languages have adopted. So I think the purpose of this rule is not to warn all implicit type conversions. In fact, this proposal is warning to only *explicit* type conversions with shorter notations. And its implementation is using implicit type conversions.

I thought the purpose of this rule is to warn confusable notations to beginners for *explicit* type conversions and to suggest more self-explanatory notations for *explicit* type conversions.

So my thought of the name is `no-(cool-but-confusable)-cast`, not `no-implicit-cast`.
I felt `tricky` matches it.
But I'm not used to English, I want advice.

My other ideas: `no-cryptic-cast`, `no-symbolic-cast`, `no-short-explicit-cast`, or `no-confusable-cast`.

**Comment by nzakas** [comment]
How about `no-implicit-coercion`?

**Comment by erikmueller** [comment]
Since the jscs rule says `disallow-implicit-type-conversion` and coercion pretty precisely describes this type conversion I'd also prefer `no-implicit-coercion`. Nonetheless people not knowing coercion might be confused and would expect sth like the `type-conversion` (especially non english-speaking people I think). On the other hand _implicit_ is kind of included in the term _coercion_ so maybe there's a better word when going with the _coercion_ term.

**Comment by mysticatea** [comment]
Wow, I like `no-implicit-coercion`.
I could image it is explicit type conversions with using implicit type conversions.

**Comment by mysticatea** [comment]
Thanks a lot for many advice.
